### Тема «Создание признакового пространства»

Продолжим обработку данных с Твиттера

In [1]:
import pandas as pd
import numpy as np
import warnings 
warnings.filterwarnings("ignore")
from sklearn import model_selection, preprocessing, linear_model
from sklearn.metrics import accuracy_score

In [2]:
df = pd.read_pickle("./result.pkl")
df.head()

id  label                                              tweet  \
0   1    0.0   @user when a father is dysfunctional and is s...   
1   2    0.0  @user @user thanks for #lyft credit i can't us...   
2   3    0.0                                bihday your majesty   
3   4    0.0  #model   i love u take with u all the time in ...   
4   5    0.0             factsguide: society now    #motivation   

                                         clean_tweet  \
0  when father is dysfunctional and is so selfish...   
1  thanks for lyft credit cannot use cause they d...   
2                                bihday your majesty   
3    model love you take with you all the time in ur   
4                  factsguide society now motivation   

                                         tweet_token  \
0  [when, father, is, dysfunctional, and, is, so,...   
1  [thanks, for, lyft, credit, can, not, use, cau...   
2                            [bihday, your, majesty]   
3  [model, love, you, take, with, you, all, the, ...   
4             [factsguide, society, now, motivation]   

                                tweet_token_filtered  \
0  [father, dysfunctional, selfish, drags, kids, ...   
1  [thanks, lyft, credit, use, cause, offer, whee...   
2                                  [bihday, majesty]   
3                      [model, love, take, time, ur]   
4                  [factsguide, society, motivation]   

                                       tweet_stemmed  \
0  [father, dysfunct, selfish, drag, kid, dysfunc...   
1  [thank, lyft, credit, use, caus, offer, wheelc...   
2                                  [bihday, majesti]   
3                      [model, love, take, time, ur]   
4                        [factsguid, societi, motiv]   

                                    tweet_lemmatized  
0  [father, dysfunctional, selfish, drag, kid, dy...  
1  [thanks, lyft, credit, use, cause, offer, whee...  
2                                  [bihday, majesty]  
3                      [model, love, take, time, ur]  
4                  [factsguide, society, motivation]

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

● Игнорируем слова, частота которых в документе строго превышает порог 0.9 с
помощью max_df.

● Ограничим количество слов, попадающий в мешок, с помощью max_features =
1000.

● Исключим стоп-слова с помощью stop_words='english'.

● Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с
помощью CountVectorizer.get_feature_names().

In [3]:
from sklearn.feature_extraction.text import CountVectorizer

count_vectorizer = CountVectorizer(max_df = 0.9, max_features = 1000, stop_words='english')

tweet_stemmed = [' '.join(tweet) for tweet in df['tweet_stemmed'].to_list()]
tweet_lemmatized = [' '.join(tweet) for tweet in df['tweet_lemmatized'].to_list()]

# Создаем the Bag-of-Words модель
bow_stemmed = count_vectorizer.fit_transform(tweet_stemmed)
bow_lemmatized = count_vectorizer.fit_transform(tweet_lemmatized)

In [4]:
# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(bow_stemmed.toarray(), columns = feature_names)

able  absolutely  account  act  action  actor  actually  adapt  add  \
0         0           0        0    0       0      0         0      0    0   
1         0           0        0    0       0      0         0      0    0   
2         0           0        0    0       0      0         0      0    0   
3         0           0        0    0       0      0         0      0    0   
4         0           0        0    0       0      0         0      0    0   
...     ...         ...      ...  ...     ...    ...       ...    ...  ...   
49154     0           0        0    0       0      0         0      0    0   
49155     0           0        0    0       0      0         0      0    0   
49156     0           0        0    0       0      0         0      0    0   
49157     0           0        0    0       0      0         0      0    0   
49158     0           0        0    0       0      0         0      0    0   

       adventure  ...  yes  yesterday  yo  yoga  york  young  youth  youtube  \
0              0  ...    0          0   0     0     0      0      0        0   
1              0  ...    0          0   0     0     0      0      0        0   
2              0  ...    0          0   0     0     0      0      0        0   
3              0  ...    0          0   0     0     0      0      0        0   
4              0  ...    0          0   0     0     0      0      0        0   
...          ...  ...  ...        ...  ..   ...   ...    ...    ...      ...   
49154          0  ...    0          0   0     0     0      0      0        0   
49155          0  ...    0          0   0     0     0      0      0        0   
49156          0  ...    0          0   0     0     0      0      0        0   
49157          0  ...    0          0   0     0     0      0      0        0   
49158          0  ...    0          0   0     0     0      0      0        0   

       yr  yummy  
0       0      0  
1       0      0  
2       0      0  
3       0      0  
4       0      0  
...    ..    ...  
49154   0      0  
49155   0      0  
49156   0      0  
49157   0      0  
49158   0      0  

[49159 rows x 1000 columns]

In [5]:
pd.DataFrame(bow_lemmatized.toarray(), columns = feature_names)

able  absolutely  account  act  action  actor  actually  adapt  add  \
0         0           0        0    0       0      0         0      0    0   
1         0           0        0    0       0      0         0      0    0   
2         0           0        0    0       0      0         0      0    0   
3         0           0        0    0       0      0         0      0    0   
4         0           0        0    0       0      0         0      0    0   
...     ...         ...      ...  ...     ...    ...       ...    ...  ...   
49154     0           0        0    0       0      0         0      0    0   
49155     0           0        0    0       0      0         0      0    0   
49156     0           0        0    0       0      0         0      0    0   
49157     0           0        0    0       0      0         0      0    0   
49158     0           0        0    0       0      0         0      0    0   

       adventure  ...  yes  yesterday  yo  yoga  york  young  youth  youtube  \
0              0  ...    0          0   0     0     0      0      0        0   
1              0  ...    0          0   0     0     0      0      0        0   
2              0  ...    0          0   0     0     0      0      0        0   
3              0  ...    0          0   0     0     0      0      0        0   
4              0  ...    0          0   0     0     0      0      0        0   
...          ...  ...  ...        ...  ..   ...   ...    ...    ...      ...   
49154          0  ...    0          0   0     0     0      0      0        0   
49155          0  ...    0          0   0     0     0      0      0        0   
49156          0  ...    0          0   0     0     0      0      0        0   
49157          0  ...    0          0   0     0     0      0      0        0   
49158          0  ...    0          0   0     0     0      0      0        0   

       yr  yummy  
0       0      0  
1       0      0  
2       0      0  
3       0      0  
4       0      0  
...    ..    ...  
49154   0      0  
49155   0      0  
49156   0      0  
49157   0      0  
49158   0      0  

[49159 rows x 1000 columns]

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.

● Игнорируем слова, частота которых в документе строго превышает порог 0.9 с
помощью max_df.

● Ограничим количество слов, попадающий в мешок, с помощью max_features =
1000.

● Исключим стоп-слова с помощью stop_words='english'.

● Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с
помощью TfidfVectorizer.get_feature_names().

In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(max_df = 0.9, max_features = 1000, stop_words='english')

values_stemmed = tfidf_vectorizer.fit_transform(tweet_stemmed)
values_lemmatized = tfidf_vectorizer.fit_transform(tweet_lemmatized)

In [7]:
# Отобразим модель как pandas DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(values_stemmed.toarray(), columns = feature_names)

able  absolutely  account  act  action  actor  actually  adapt  add  \
0       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
1       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
2       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
3       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
4       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
...     ...         ...      ...  ...     ...    ...       ...    ...  ...   
49154   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49155   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49156   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49157   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49158   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   

       adventure  ...  yes  yesterday   yo  yoga  york  young  youth  youtube  \
0            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
1            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
2            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
3            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
4            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
...          ...  ...  ...        ...  ...   ...   ...    ...    ...      ...   
49154        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49155        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49156        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49157        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49158        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   

        yr  yummy  
0      0.0    0.0  
1      0.0    0.0  
2      0.0    0.0  
3      0.0    0.0  
4      0.0    0.0  
...    ...    ...  
49154  0.0    0.0  
49155  0.0    0.0  
49156  0.0    0.0  
49157  0.0    0.0  
49158  0.0    0.0  

[49159 rows x 1000 columns]

In [8]:
pd.DataFrame(values_lemmatized.toarray(), columns = feature_names)

able  absolutely  account  act  action  actor  actually  adapt  add  \
0       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
1       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
2       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
3       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
4       0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
...     ...         ...      ...  ...     ...    ...       ...    ...  ...   
49154   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49155   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49156   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49157   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   
49158   0.0         0.0      0.0  0.0     0.0    0.0       0.0    0.0  0.0   

       adventure  ...  yes  yesterday   yo  yoga  york  young  youth  youtube  \
0            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
1            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
2            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
3            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
4            0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
...          ...  ...  ...        ...  ...   ...   ...    ...    ...      ...   
49154        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49155        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49156        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49157        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   
49158        0.0  ...  0.0        0.0  0.0   0.0   0.0    0.0    0.0      0.0   

        yr  yummy  
0      0.0    0.0  
1      0.0    0.0  
2      0.0    0.0  
3      0.0    0.0  
4      0.0    0.0  
...    ...    ...  
49154  0.0    0.0  
49155  0.0    0.0  
49156  0.0    0.0  
49157  0.0    0.0  
49158  0.0    0.0  

[49159 rows x 1000 columns]

3. Проверьте ваши векторайзеры на корпусе который использовали на вебинаре, составьте таблицу метод векторизации и скор который вы получили (в методах векторизации по изменяйте параметры что бы добиться лучшего скора) обратите внимание как падает/растёт скор при уменьшении количества фичей, и изменении параметров, так же попробуйте применить к векторайзерам PCA для сокращения размерности посмотрите на качество сделайте выводы

In [9]:
# Загружаем данные
data = open('corpus').read()
labels, texts = [], []
for i, line in enumerate(data.split("\n")):
    content = line.split()
    labels.append(content[0])
    texts.append(" ".join(content[1:]))

# создаем df
trainDF = pd.DataFrame()
trainDF['text'] = texts
trainDF['label'] = labels
trainDF.head(5)

text       label
0  Stuning even for the non-gamer: This sound tra...  __label__2
1  The best soundtrack ever to anything.: I'm rea...  __label__2
2  Amazing!: This soundtrack is my favorite music...  __label__2
3  Excellent Soundtrack: I truly like this soundt...  __label__2
4  Remember, Pull Your Jaw Off The Floor After He...  __label__2

In [10]:
train_x, valid_x, train_y, valid_y = model_selection.train_test_split(trainDF['text'], trainDF['label'], random_state=24)

# labelEncode целевую переменную
encoder = preprocessing.LabelEncoder()
train_y = encoder.fit_transform(train_y)
valid_y = encoder.fit_transform(valid_y)

count_vectorizer.fit(trainDF['text'])

xtrain_count = count_vectorizer.transform(train_x)
xvalid_count = count_vectorizer.transform(valid_x)

classifier = linear_model.LogisticRegression()
classifier.fit(xtrain_count, train_y)
pred_count = classifier.predict(xvalid_count)
pred_count

array([0, 1, 0, ..., 0, 0, 1])

In [11]:
tfidf_vectorizer.fit(trainDF['text'])

xtrain_tfidf = tfidf_vectorizer.transform(train_x)
xvalid_tfidf = tfidf_vectorizer.transform(valid_x)

classifier.fit(xtrain_tfidf, train_y)
pred_tfidf = classifier.predict(xvalid_tfidf)
pred_tfidf

array([0, 1, 0, ..., 0, 0, 1])

In [12]:
d = {'count_vectorizer': accuracy_score(valid_y, pred_count), 'tfidf_vectorizer': accuracy_score(valid_y, pred_tfidf)}
acc_df = pd.DataFrame(data=d, index=['accuracy'])
acc_df

count_vectorizer  tfidf_vectorizer
accuracy             0.826            0.8312

In [13]:
for n in range(1,11):
    count_vect = CountVectorizer(max_df = 0.9, max_features = n*500, stop_words='english')
    tfidf_vect = TfidfVectorizer(max_df = 0.9, max_features = n*500, stop_words='english')

    count_vect.fit(trainDF['text'])
    tfidf_vect.fit(trainDF['text'])

    xtrain_count1 = count_vect.transform(train_x)
    xvalid_count1 = count_vect.transform(valid_x)
    xtrain_tfidf1 = tfidf_vect.transform(train_x)
    xvalid_tfidf1 = tfidf_vect.transform(valid_x)

    classifier.fit(xtrain_count1, train_y)
    pred_count1 = classifier.predict(xvalid_count1)
    
    classifier.fit(xtrain_tfidf1, train_y)
    pred_tfidf1 = classifier.predict(xvalid_tfidf1)

    print(f'max_features: {n*500}, \
            count_accuracy: {accuracy_score(valid_y, pred_count1)}, \
            tfidf_accuracy: {accuracy_score(valid_y, pred_tfidf1)}')

max_features: 500,             count_accuracy: 0.8128,             tfidf_accuracy: 0.8124
max_features: 1000,             count_accuracy: 0.826,             tfidf_accuracy: 0.8312
max_features: 1500,             count_accuracy: 0.828,             tfidf_accuracy: 0.84
max_features: 2000,             count_accuracy: 0.8292,             tfidf_accuracy: 0.8412
max_features: 2500,             count_accuracy: 0.8292,             tfidf_accuracy: 0.8488
max_features: 3000,             count_accuracy: 0.8324,             tfidf_accuracy: 0.8516
max_features: 3500,             count_accuracy: 0.8344,             tfidf_accuracy: 0.852
max_features: 4000,             count_accuracy: 0.8336,             tfidf_accuracy: 0.852
max_features: 4500,             count_accuracy: 0.836,             tfidf_accuracy: 0.85
max_features: 5000,             count_accuracy: 0.8348,             tfidf_accuracy: 0.8524


Видим, что при увеличении максимального кол-ва фичей скор также растет.

3. Натренируем gensim.models.Word2Vec модель на наших данных.
Тренировать будем на токенизированных твитах combine_df['tweet_token']
Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

In [14]:
# pip install gensim

In [15]:
from gensim.models import Word2Vec

In [16]:
modelW2V = Word2Vec(sentences=df['tweet_token'], 
                    vector_size=200, 
                    window=5, 
                    min_count=2, 
                    sg = 1, 
                    hs = 0, 
                    negative = 10, 
                    workers= 32, 
                    seed = 34)

In [17]:
modelW2V.train(df['tweet_token'],
               total_examples = df['tweet_token'].size, 
               epochs = 20)

(9142015, 11726520)

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [18]:
modelW2V.wv.most_similar(positive=['dinner'])

[('bolognese', 0.5685358047485352),
 ('spaghetti', 0.5549445152282715),
 ('bihdaydinner', 0.5541219711303711),
 ('shawarma', 0.5539870858192444),
 ('cookout', 0.5467182397842407),
 ('hamburger', 0.5413327813148499),
 ('tacotuesday', 0.5381819009780884),
 ('burritos', 0.5291736721992493),
 ('sizzle', 0.5225168466567993),
 ('waterloo', 0.5215371251106262)]

In [19]:
modelW2V.wv.most_similar(positive=['trump'])

[('donald', 0.5734719038009644),
 ('impeachment', 0.5165001749992371),
 ('fuhered', 0.5163795351982117),
 ('dumptrump', 0.5140902996063232),
 ('donaldtrump', 0.5073265433311462),
 ('delegaterevolt', 0.5031085014343262),
 ('presidentelect', 0.49981024861335754),
 ('crony', 0.49849212169647217),
 ('suppoer', 0.4976588189601898),
 ('trumptrain', 0.49590152502059937)]

5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

In [20]:
modelW2V.wv['food']

array([ 0.4258697 , -0.5026537 , -0.15994477, -0.00442187, -1.1539103 ,
       -0.67788935,  0.19382793,  0.31653154, -0.5208238 ,  0.20689003,
       -0.15577476,  0.32235998,  0.22695518, -0.047918  , -0.7127395 ,
        0.35528818, -0.10383546, -0.5754716 , -0.32809168, -0.2126159 ,
       -0.5924227 , -0.18656313, -0.39002544, -0.2382206 ,  0.21852657,
        0.25594908, -0.11054653, -0.3671149 ,  0.63977706, -0.2616428 ,
        1.2648827 ,  0.15992896, -0.41397503, -0.541881  , -0.3870668 ,
       -0.80563384, -0.17072642,  0.12514918,  0.12573421, -0.42109922,
        0.15108566, -0.18695208,  0.0339105 ,  0.33902583, -0.05268939,
       -0.01423708, -0.25868562, -0.46088323, -1.1223211 ,  0.39632225,
        0.15117002, -0.15145919,  0.13690329, -0.60038984,  0.28863266,
       -0.00307634, -0.3986549 , -1.1800028 ,  0.01413429,  0.4045048 ,
       -0.4633804 , -0.49626383,  0.23168753, -0.5448631 , -0.02370846,
        0.43866664,  0.22011374, -0.01412236,  0.54391354, -0.79

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.

Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).


In [21]:
wordvec_df = np.zeros((len(df), 200))
for i, tweet in enumerate(df['tweet_token']):
    vec = np.zeros((1, 200))
    for word in tweet:
        try:
            vec += modelW2V.wv[word].reshape((1, 200))
        except:
            continue
    vec /= len(tweet)
    wordvec_df[i] = vec

In [22]:
wordvec_df.shape

(49159, 200)